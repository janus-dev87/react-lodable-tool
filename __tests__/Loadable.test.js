import React from "react";
import renderer from "react-test-renderer";
import Loadable from "../src";

let createComponent = (name: string) => {
  return (props: mixed) => {
    return <div>{name} {JSON.stringify(props)}</div>;
  };
};

let waitFor = (delay: number) => {
  return new Promise(resolve => {
    setTimeout(resolve, delay);
  });
};

let createLoader = (
  delay: number,
  Component: React.Component | false,
  error?: error
) => {
  return async () => {
    await waitFor(delay);
    if (Component) {
      return Component;
    } else {
      throw error;
    }
  };
};

let LoadingComponent = createComponent("LoadingComponent");
let ErrorComponent = createComponent("ErrorComponent");
let MyComponent = createComponent("MyComponent");

test("loading success", async () => {
  let LoadableMyComponent = Loadable(
    createLoader(400, LoadingComponent),
    LoadingComponent
  );

  let component1 = renderer.create(<LoadableMyComponent prop="foo" />);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loaded

  let component2 = renderer.create(<LoadableMyComponent prop="bar" />);

  expect(component2.toJSON()).toMatchSnapshot(); // reload
});

test("loading error", async () => {
  let LoadableMyComponent = Loadable(
    createLoader(400, null, new Error("test error")),
    LoadingComponent,
    ErrorComponent
  );

  let component1 = renderer.create(<LoadableMyComponent prop="baz" />);

  expect(component1.toJSON()).toMatchSnapshot(); // initial
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // loading
  await waitFor(200);
  expect(component1.toJSON()).toMatchSnapshot(); // errored
});
