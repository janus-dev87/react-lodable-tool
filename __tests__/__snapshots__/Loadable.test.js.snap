// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading error 1`] = `null`;

exports[`loading error 2`] = `
<div>
  LoadingComponent
   
  {}
</div>
`;

exports[`loading error 3`] = `
<div>
  ErrorComponent
   
  {"error":{}}
</div>
`;

exports[`loading success 1`] = `null`;

exports[`loading success 2`] = `
<div>
  LoadingComponent
   
  {}
</div>
`;

exports[`loading success 3`] = `
<div>
  LoadingComponent
   
  {"prop":"foo"}
</div>
`;

exports[`loading success 4`] = `
<div>
  LoadingComponent
   
  {"prop":"bar"}
</div>
`;
